command/job_restart.go

```diff
@@ -0,0 +1,1185 @@
+//
+// The flags -batch-wait and -on-error have an 'ask' option. This function
+// handles both to prevent asking the user twice in case they are both set to
+// 'ask' and an error happens.
+func (c *JobRestartCommand) shouldProceed(err error) bool {
+	var question, options string
+
+	if err == nil {
+		question = "Proceed with the next batch?"
+		options = "Y/n"
+	} else {
+		question = "Ignore the errors above and proceed with the next batch?"
+		options = "y/N" // Defaults to 'no' if an error happens.
+
+		if !c.isErrorRecoverable(err) {
+			question = `The errors above are likely to happen again.
+Ignore them anyway and proceed with the next batch?`
+		}
+	}
+
+	// If -batch-wait is 'ask' the user can provide a new wait duration.
+	if c.batchWaitAsk {
+		options += "/<wait duration>"
+	}
+
+	return c.askQuestion(
+		fmt.Sprintf("%s [%s]", question, options),
+		false,
+		func(answer string) (bool, error) {
+			switch strings.ToLower(answer) {
+			case "":
+				// Proceed by default only if there is no error.
+				return err == nil, nil
+			case "y", "yes":
+				return true, nil
+			case "n", "no":
+				return false, nil
+			default:
+				if c.batchWaitAsk {
+					// Check if user passed a time duration and adjust the
+					// command to use that moving forward.
+					batchWait, err := time.ParseDuration(answer)
+					if err != nil {
+						return false, fmt.Errorf("Invalid answer %q", answer)
+					}
+
+					c.batchWaitAsk = false
+					c.batchWait = batchWait
+					c.Ui.Output(c.Colorize().Color(fmt.Sprintf(
+						"[bold]==> %s: Proceeding restarts with new wait time of %s[reset]",
+						formatTime(time.Now()),
+						c.batchWait,
+					)))
+					return true, nil
+				} else {
+					return false, fmt.Errorf("Invalid answer %q", answer)
+				}
+			}
+		})
+}
+
+// shouldExit blocks and waits for the user for confirmation if they would like
+// to interrupt the command. Returns true if the answer is positive.
+func (c *JobRestartCommand) shouldExit() bool {
+	question := `Restart interrupted, no more allocations will be restarted.
+Are you sure you want to stop the restart process? [y/N]`
+
+	return c.askQuestion(
+		question,
+		true,
+		func(answer string) (bool, error) {
+			switch strings.ToLower(answer) {
+			case "n", "no", "":
+				return false, nil
+			case "y", "yes":
+				return true, nil
+			default:
+				return false, fmt.Errorf("Invalid answer %q", answer)
+			}
+		})
+}
+
+// askQuestion asks question to user until they provide a valid response.
+func (c *JobRestartCommand) askQuestion(question string, onError bool, cb func(string) (bool, error)) bool {
+	prefixedQuestion := fmt.Sprintf(
+		"[bold]==> %s: %s[reset]",
+		formatTime(time.Now()),
+		indentString(question, jobRestartTimestampPrefixLength),
+	)
+
+	// Let ui.Ask() handle interrupt signals.
+	signal.Stop(c.sigsCh)
+	defer func() {
+		signal.Notify(c.sigsCh, syscall.SIGINT)
+	}()
+
+	for {
+		answer, err := c.Ui.Ask(c.Colorize().Color(prefixedQuestion))
```

Comment:
While doing some testing I noticed that `Ask` has some weird behavior around buffering where some errant keypresses after `Ctrl-C` got dropped into the input. I'm not sure if we can actually do anything about it here in Nomad, but just leaving a note in case we encounter it again:

```
    2023-03-17T10:06:07-04:00: Still waiting for allocation "1b86450c" to be replaced
^C^[[A^[[A^[^[==> 2023-03-17T10:06:13-04:00: Restart interrupted, no more allocations will be restarted.
                               Are you sure you want to stop the restart process? [y/N] y
                               Invalid answer "\x1b[A\x1b[A\x1b\x1by"
==> 2023-03-17T10:06:13-04:00: Restart interrupted, no more allocations will be restarted.
                               Are you sure you want to stop the restart process? [y/N] y
```

Reply:
Ah interesting! We could try to remove signal characters from the input `Ask` gives us, but the list of signals and their values may change from system to system. 

It may also be possible to [`Ignore`](https://pkg.go.dev/os/signal#Ignore) all other signals while an input is being provided, but that can be pretty annoying since it breaks the expected flow of signal handling.

All of this to say that I agree there's not much we can do here...